### PR TITLE
Adds coco access token lifecycle

### PR DESCRIPTION
**Pull request by deekshith-elear**
- Adds base code structure for application.
- Adds code for getting access token using clientID and clientsecret

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  },
+  "author": "",
+  "bugs": {
+    "url": "https://github.com/ashishbajaj99/nodesampletokenserver/issues"
+  },
+  "homepage": "https://github.com/ashishbajaj99/nodesampletokenserver#readme",
+  "dependencies": {
+    "body-parser": "^1.18.2",
+    "cookie-parser": "^1.4.4",
+    "elear-node-libs": "git+https://68add7737f50e20e177b611b288e0eb78bab3654:x-oauth-basic@github.com/ashishbajaj99/elear-node-libs.git#develop",
```
Comment (agarwal13) [inline package.json]: - We need to check if we will be exposing this. I dont think so. So, in the end, we need to refactor the code to remove the elear-node-libs from the dependency list. We can confirm with Ashish once on this

File: package.json [hunk 2]
```diff
@@ -0,0 +1,45 @@
+{
+  "name": "nodesampletokenserver",
+  "version": "1.0.0",
+  "description": "Sample application to get access tokens of non coco users using service apllication in coco",
```
Comment (agarwal13) [inline package.json]: We need to update the description appropriately. This is sample service application of COCO to allow non coco users to access the COCONet functionality

File: server/api/v1.0/coco-api/index.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  * merged, published, distributed, sublicensed, reproduced in whole or in part,
+  * nor its contents revealed in any manner to others without the express
+  * written permission of Elear Solutions Tech Private Limited.
+  */
+/*********************************************************************************/
+/*===============================================================================*/
+
+import express from 'express';
+
+// internal network router
```
Comment (agarwal13) [inline server/api/v1.0/coco-api/index.js]: - correct the comment

File: server/config/coco.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        grant_type: COCO_SERVICE_APPLICATION_GRANT_TYPE,
+        client_id: Environment.COCO_CLIENT_ID,
+        client_secret: Environment.COCO_CLIENT_SECRET
+      },
+      json: true,
+    }, (err, response, body) => {
+      if (err || response.statusCode !== HttpStatus.OK) {
+        logger.error('initialize: Error occurred while'
+          + ' fetching access token', err);
+        return reject([err, response.statusCode, body]);
```
Comment (agarwal13) [inline server/config/coco.js]: - Why are we using this format here?

File: server/utils/constants.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+// service name to be used in logs
+export const SERVICE_NAME = "SAMPLETOKENSERVER";
+
+// COCO confedential client auth grantType
+export const COCO_SERVICE_APPLICATION_GRANT_TYPE = "client_credentials";
+
+// COCO refresh token auth grantType
+export const COCO_REFRESH_TOKEN_GRANT_TYPE = "refresh_token";
+
```
Comment (agarwal13) [inline server/utils/constants.js]: - Create an Object called OAuthGrantTypes and inside that create keys like CLIENT_CREDENTIALS and REFRESH_TOKEN

File: server/config/coco.js [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    .then(() => {
+      // Check if access token is expired,
+      // if true refresh access token
+      if (COCOTokenInfo.expiryTime - timestamp <= AUTH_TOKEN_TIMEOUT_BUFFER) {
+        return refreshTokens();
+      }
+
+      return Promise.resolve();
+    }).then(() => {
+      return Promise.resolve(COCOTokenInfo.accessToken);
```
Comment (agarwal13) [inline server/config/coco.js:129]: - i think its fine to just return COCOTokenInfo.accessToken here

File: server/config/common-config.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+ * recipient agrees that this material and the information contained therein is
+ * held in confidence and in trust and will NOT be used, copied, modified,
+ * merged, published, distributed, sublicensed, reproduced in whole or in part,
+ * nor its contents revealed in any manner to others without the express
+ * written permission of Elear Solutions Tech Private Limited.
+ */
+/*********************************************************************************/
+/*===============================================================================*/
+
+import common from 'elear-node-libs/coco-store-common';
```
Comment (agarwal13) [inline server/config/common-config.js]: - this file depends on if we are allowed to use this library, which i dont think so

File: server/routes.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+/*********************************************************************************/
+/*===============================================================================*/
+
+import { logger } from './config/common-config';
+import { HttpStatus } from './utils/custom-api-errors';
+import { cocoAPIRouter } from './api/v1.0/coco-api';
+
+// Routes function to load apis on specific routes
+export default (app) => {
+  app.use('/v1.0/coco-api', cocoAPIRouter);
```
Comment (agarwal13) [inline server/routes.js]: - coco-api doest look to be appropriate. lets discuss

File: server/server.js [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+app.use(bodyParser.urlencoded({ extended: false }));
+
+// parse cookie header and populate req.cookies
+app.use(cookieParser());
+
+// compression middleware
+app.use(shrinkRay());
+
+// webapp security middleware
+app.use(lusca(security.Lusca));
```
Comment (agarwal13) [inline server/server.js]: - this security.Lusca also depends on the inclusion of the elear-node-libs.

File: server/server.js [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+router(app);
+
+// error-handler to be used last and in dev only
+if (process.env.NODE_ENV === 'dev') {
+  app.use(errorHandler());
+}
+// Catch all other routes and return the NOT_FOUND
+app.get('*', (req, res) => {
+  res.status(HttpStatus.NOT_FOUND).send();
+});
```
Comment (agarwal13) [inline server/server.js]: - is this not redundant as we already have the * route in the router.js?

File: server/utils/custom-api-errors.js [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+// Export CustomApiError and HttpStatus
+export const CustomApiError = error.CustomApiError;
+export const HttpStatus = error.HttpStatus;
+
+/** List of Custom API ErrorCodes */
+export const CustomErrorCodes  = {
+};
+
+// Export Error Codes
+export const ErrorCodes = error.ErrorCodes(CustomErrorCodes);
```
Comment (agarwal13) [inline server/utils/custom-api-errors.js]: - Even this file needs to be relooked into

**Review by agarwal13 — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (deekshith-elear) [reply package.json]: As, per discussion with ashish we no need to expose this, we have to keep error codes in constants and for logs use console.log

File: package.json (hunk 2 above)
Comment (deekshith-elear) [reply package.json]: handled

File: server/api/v1.0/coco-api/index.js (hunk 3 above)
Comment (deekshith-elear) [reply server/api/v1.0/coco-api/index.js]: handled

File: server/utils/constants.js (hunk 5 above)
Comment (deekshith-elear) [reply server/utils/constants.js]: handled

File: server/config/common-config.js (hunk 7 above)
Comment (deekshith-elear) [reply server/config/common-config.js]: yes these will be removed once all PR's are done

File: server/utils/custom-api-errors.js (hunk 11 above)
Comment (deekshith-elear) [reply server/utils/custom-api-errors.js]: these will be removed as discussed

File: server/server.js (hunk 10 above)
Comment (deekshith-elear) [reply server/server.js]: removed

File: server/server.js (hunk 9 above)
Comment (deekshith-elear) [reply server/server.js]: removed

File: server/routes.js (hunk 8 above)
Comment (deekshith-elear) [reply server/routes.js]: This will be updating at last

**Review by sohinihm4 — APPROVED** [review]

**Review by sohinihm4 — APPROVED** [review]

**Pull request merged**